### ISSUE TITLE
Customize segmentation model architecture 

### ISSUE THREAD

**Issue by sadimoodi** [ultralytics/ultralytics#2562]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Thank you so much guys for the outstanding work!
@glenn-jocher I know that building larger models to support image size = 1280 is in your road map, however, i do not want to wait because i do not plan to transfer weights from your pretrained models. (plan to train from scratch).
The current segmentation model has 71,752774 parameters, I plan to reach 500M.
Looking at "yolov8-seg.yaml" i do not seem to understand how to add layers in YOLOv8, are there any best practices i can follow?
Aprpeciate if you provide any guidance on how to build a larger model.
Thanks again

**Comment by glenn-jocher** [comment]
@sadimoodi thank you for your kind words and interest in YOLOv8! 

Regarding your question about adding layers to YOLOv8, you can follow the steps below:

1. Create a new `convolutional` layer in the model config (e.g. `yolov8-seg.yaml`) with appropriate `filters`, `size` and `stride` parameters.
2. Add this new layer to the end of the layer list in the `head` section (after the previous `convolutional` layer).
3. Update the following layers in the `head` section with regards to their `filters` parameter. For example, if you add a new layer with 256 `filters`, layer `6` should have 512 filters, layer `9` should have 256 filters, and layer `12` should have 128 filters. 
4. Make sure to update the number of classes in the `head` section.

We hope this helps! Please let us know if you have further questions.

**Comment by sadimoodi** [comment]
@glenn-jocher  Thank you for your reply, i am a little confused of the instructions mentioned above, here is the original head in "yolov8-seg.yaml": 
# YOLOv8.0n head
head:
  - [-1, 1, nn.Upsample, [None, 2, 'nearest']]
  - [[-1, 6], 1, Concat, [1]]  # cat backbone P4
  - [-1, 3, C2f, [512]]  # 12

  - [-1, 1, nn.Upsample, [None, 2, 'nearest']]
  - [[-1, 4], 1, Concat, [1]]  # cat backbone P3
  - [-1, 3, C2f, [256]]  # 15 (P3/8-small)

  - [-1, 1, Conv, [256, 3, 2]]
  - [[-1, 12], 1, Concat, [1]]  # cat head P4
  - [-1, 3, C2f, [512]]  # 18 (P4/16-medium)

  - [-1, 1, Conv, [512, 3, 2]]
  - [[-1, 9], 1, Concat, [1]]  # cat head P5
  - [-1, 3, C2f, [1024]]  # 21 (P5/32-large)

  - [[15, 18, 21], 1, Segment, [nc, 32, 256]]  # Segment(P3, P4, P5)

Here is my modified head as i understood from your suggestion:
# YOLOv8.0n NEW head
head:
  - [-1, 1, nn.Upsample, [None, 2, 'nearest']]
  - [[-1, 6], 1, Concat, [1]]  # cat backbone P4
  - [-1, 3, C2f, [512]]  # 12

  - [-1, 1, nn.Upsample, [None, 2, 'nearest']]
  - [[-1, 4], 1, Concat, [1]]  # cat backbone P3
  - [-1, 3, C2f, [256]]  # 15 (P3/8-small)

  - [-1, 1, Conv, [256, 3, 2]]
  - [[-1, 12], 1, Concat, [1]]  # cat head P4
  - [-1, 3, C2f, [512]]  # 18 (P4/16-medium)

  - [-1, 1, Conv, [512, 3, 2]]
  - [[-1, 9], 1, Concat, [1]]  # cat head P5
  - [-1, 3, C2f, [1024]]  # 21 (P5/32-large)

  - [-1, 1, Conv, [256, 3, 2]]
  - [[-1, 12], 1, Concat, [1]]  # cat head P6
  - [-1, 3, C2f, [512]]  # 24 (P4/16-medium)1

  - [-1, 1, Conv, [512, 3, 2]]
  - [[-1, 9], 1, Concat, [1]]  # cat head P7
  - [-1, 3, C2f, [1024]]  # 27 (P5/32-large)

  - [-1, 1, Conv, [256, 3, 2]]
  - [[-1, 12], 1, Concat, [1]]  # cat head P8
  - [-1, 3, C2f, [512]]  # 30 (P4/16-medium)1

  - [-1, 1, Conv, [128, 3, 2]]
  - [[-1, 12], 1, Concat, [1]]  # cat head P9
  - [-1, 3, C2f, [512]]  # 33 (P4/16-medium)1

  - [[15, 18, 21], 1, Segment, [nc, 32, 256]]  # Segment(P3, P4, P5)

Would the above suggested head work during training and inference and increase the number of params significantly? 
Also what about the backbone section? do i need to modify it as well?
I would like to get a model that is at least 5 times larger than the current model in terms of number of params.
it would be very helpfull if you share the new suggested "yolov8-seg.yaml" file.
Thank again for your wonderful work.

**Comment by sadimoodi** [comment]
@glenn-jocher i am facing errors, appreciate if u share the head and backbone code sections after adding new layers to both of them

**Comment by glenn-jocher** [comment]
@sadimoodi if you want to create a P6 segmentation model the simplest way is to start from yolov8-p6.yaml and then just replace Detect() with Segment().

**Comment by sadimoodi** [comment]
Thanks @glenn-jocher for your reply.
in the yaml file, there is this last line of code:
` - [[15, 18, 21], 1, Segment, [nc, 32, 256]] # Segment(P3, P4, P5)`

can you explain what do 32, 256 represent? appreciate if you shed some light on whether increasing these values might improve segmentation mask? also does 1 mean that there is only 1 mask? if so, would increasing this number result in a better mask?

**Comment by glenn-jocher** [comment]
@sadimoodi certainly! The `[nc, 32, 256]` argument in the `Segment` layer of the `yolov8-segment.yaml` file specifies the parameters for the final segmentation layer of the network. The three values correspond to:

- `nc`: the number of classes that the network is trained to segment. This value should be set according to the number of different object categories in your segmentation dataset. For example, if your dataset contains three categories (e.g., "sky", "tree", "person"), you should set `nc` to `3`.
- `32`: this represents the number of channels in the final feature maps generated by the segmentation head. Increasing this value may improve segmentation accuracy but will also increase the number of parameters in the network. You can experiment with different values of this parameter to find the optimal configuration for your specific segmentation task.
- `256`: this value represents the size of the upscaled segmentation mask. The size of the output mask can be increased by increasing this value, but doing so will also increase the amount of memory required to store the network output. Similarly to the previous parameter, you can experiment with different values to find the optimal size for your task.

The `1` in `[15, 18, 21], 1, Segment, [nc, 32, 256]` means that there will be one mask generated for each input image. Increasing this value will not result in better segmentation; rather, it would be useful if you had multiple objects in the same image that needed to be segmented. If you had two objects in an image and only specified one output mask, the two objects would be merged into one segmented area, which may not be what you want.

I hope this helps! Let us know if you have any further questions.

**Comment by RyanPham19092002** [comment]
@glenn-jocher 
Hello ! I have searched the YOLO_NAS with task instance segmentation. I researched in the google, but I just find YOLO_NAS combine with SAM model. However, I want to fine tuning YOLO_NAS model which has branch mask , such as Mask RCNN, to instance segmentation output. Can you help me to suggest how to do this task ? 
Thank you a lot!

**Comment by glenn-jocher** [comment]
@RyanPham19092002 hello! Great to hear you're exploring YOLO with instance segmentation. While traditional YOLO models use bounding boxes for object detection, integrating a branch for mask generation similar to Mask RCNN would indeed allow for instance segmentation output. It’s a unique and complex task, since it combines aspects of semantic segmentation and object detection.

You would first need modify the YOLO architecture to incorporate a segmentation branch. This addition should be made in parallel to the existing layers responsible for object detection, as you would want the model to output both bounding boxes and pixel-level object masks.

Secondly, you would need to create or use a dataset that contains both bounding box annotations and pixel-level masks for each object instance, such as COCO. This data will be required for training the model.

Finally, training the new model would entail modifications to the loss function to incorporate losses from both the detection and segmentation branches.

While it is quite a complex task demanding a deep understanding of YOLO's architecture and functioning, it is a feasible project with potentially great results. Please note that an endeavor like this requires a strong background in machine learning and programming, as well as ample time for implementation and debugging.

Good luck with your project! Let us know if you have further questions.

**Comment by RyanPham19092002** [comment]
Thanks for your reply,
However, when I use function loss, I don't know the value of segmentation
loss gain, can you please tell me ? Is there parameter equal with box gain
?
        loss[0] *= self.gain_list[0]  # cls gain
        loss[1] *= self.gain_list[1]  # box gain
        loss[2] *= self.gain_list[2]  # dfl gain
        loss[3] *= self.gain_list[3] / batch_size #seg gain
Thanks you

Vào CN, 10 thg 9, 2023 vào lúc 08:01 Glenn Jocher <
***@***.***> đã viết:

**Comment by RyanPham19092002** [comment]
I just found box loss gain, cls boss gain, dfl loss gain, where I can file
seg loss gain ?
[image: image.png]
Thank you!


Vào Th 5, 28 thg 9, 2023 vào lúc 22:48 thanh phát phạm <
***@***.***> đã viết:

**Comment by glenn-jocher** [comment]
@RyanPham19092002 hello! The "seg loss gain" is a parameter defined by you and it would be used to scale the contribution of the segmentation loss in the total loss during training. Its value depends on the importance you want to give to the segmentation task compared to the other tasks (classification, bounding box prediction, etc.). 

For example, if you think predicting the segmentation mask is twice as important as predicting the bounding box, you would set the "seg loss gain" to be twice the "box gain". Conversely, if you think it's half as important, you would set it to be half the "box gain". 

Kindly bear in mind the balance between different tasks related losses, while tuning these parameters. This balance majorly influences the model's ability to learn different tasks appropriately. 

Feel free to conduct experiments by assigning different values to the "seg loss gain" to evaluate the model's performance under various settings. Good luck with your task!

**Comment by RyanPham19092002** [comment]
Thanks for your reply, your information help me a lot  !!

Vào Th 6, 29 thg 9, 2023 vào lúc 08:05 Glenn Jocher <
***@***.***> đã viết:

**Comment by glenn-jocher** [comment]
You're welcome, @RyanPham19092002! I'm glad to hear that the information was helpful for you. Keep in mind that tuning models for complex tasks like these can be a process of trial and error, so don't hesitate to continue experimenting and refining your approach based on the results you get. Wishing you all the best with your project. If you have any more questions or if there's anything else you need help with, don't hesitate to ask.

**Comment by RyanPham19092002** [comment]
I'm sorry but I got a problem when I'm training data for instance
segmentation task, I built yolo nas to solve the object detection task will
work well, but when I have a branch Proto ( like Clas Proto in YOLOv8
Segment) and config loss function (image 2 ) with self.single_mask_loss can
be calculate by image 1 . The problem is my seg_loss not change so much, it
will be approximately 7.79 (like image 3) , and then take my total_loss is
very high, how can I fix that ? Thank you
[image: image.png]
image 1
 [image: image.png]
image 2
 [image: image.png]
image 3

Vào Th 6, 29 thg 9, 2023 vào lúc 16:57 Glenn Jocher <
***@***.***> đã viết:

**Comment by glenn-jocher** [comment]
Hello @RyanPham19092002,

Since you're observing that your seg_loss doesn't change much and is staying close to 7.79, this might suggest possible issues with your training setup. Here are some things you may want to consider:

1. Learning Rate: If your learning rate is too low, it may cause your model to converge slowly, or even not at all. Conversely, a learning rate that is too high may cause the weights to diverge, again leading to no improvement.

2. Data Quality: Thoroughly check the quality and annotation of your training data. If the segmentation masks are not correctly annotated or not diverse enough, the model may not be learning effectively.

3. Augmentation: Depending on the data and the problem, sometimes applying stronger or more diverse data augmentation can help with training, as it effectively increases the size and diversity of your training set.

4. Loss Function: Validate your loss function to ensure it is properly penalising errors during training. If it's not, it might not direct the model weights towards a better solution.

5. Model Architecture: Revisit the model architecture and confirm that all layers are connected properly. A minor error in the architecture could prevent effective learning.

You might want to experiment with these aspects and monitor how each affects the `seg_loss` during training. Also, remember to monitor other metrics (e.g. validation performance) to ensure your model is indeed learning. 

Good luck!

**Comment by RyanPham19092002** [comment]
Thanks for your help, I will try this!

Vào Th 2, 2 thg 10, 2023 vào lúc 11:46 Glenn Jocher <
***@***.***> đã viết:

**Comment by glenn-jocher** [comment]
You're welcome, @RyanPham19092002! I'm glad the information was useful. Remember, debugging and fine-tuning your model can be an iterative process. It's all about experimenting with the parameters, evaluating the results and progressively getting closer to your goal. Be patient and thorough. Don't hesitate to ask if you have more questions, or need further guidance. Good luck with your YOLOv8 project!

**Comment by sadimoodi** [comment]
@glenn-jocher  i would like to approach for professional consultation regarding one project we have that is related to this topic, how to contact you?

**Comment by glenn-jocher** [comment]
@sadimoodi hello, thank you for your interest in professional consultation for your project. You can contact Glen Jocher directly through his email, which you can typically find on the author's GitHub profile or in the commit history. You can arrange for a meeting or a call to discuss your project needs further. Please provide some background of your project in your initial email for a more efficient consultation. Thanks again for your interest and we look forward to potentially working with you.

**Comment by rajas1310** [comment]
Hi @glenn-jocher, 
Thank you and your team for the amazing work! 
I have a trained a yolov8L model to detect humans in sport videos and it works very well. I aim to create a segmentation model using the body from my detection-model and the head from your pretrained segmentation model weights. I think the intermediate representations from my custom detection-model will enhance the segmentation output generated by the segmentation head.

I was able to load the detection weights from sub-modules 0 to 21 on the backbone architecture and your pretrained weights associated with the 22nd module from 'yolov8l-seg.pt'. But when I infer an image on this model, it gives masks = `None`, and empty lists for boxes attribute as well. 

The script I used to load the weights:
```
seg_model = YOLO('yolov8l-seg.pt')  # load an official model
det_model = YOLO('custom_human_detection.pt')

for i in range(22):
     seg_model.model.model[i].load_state_dict(det_model.model.model[i].state_dict())
```

The inference script:
```
results = seg_model.predict(img_path, max_det=1, verbose=False)  # predict on an image
print(results[0].masks)
p = results[0].plot(boxes=False)

try:
    if cv2.imwrite(f'image/saved/at/{im_idx}.jpg', p):
except Exception as e:
    print(e)
```
Output >>>
`None`


Could you please help me figure out what can be the problem?

**Comment by glenn-jocher** [comment]
Hi @rajas1310,

Thanks for reaching out and sharing your work with YOLOv8! It's great to hear about your success with the detection model. 😊 Regarding the issue you're encountering, it seems like there might be a mismatch or initialization problem after merging the weights from two models.

Here are a few suggestions:
- Ensure that the segmentation model architecture fully aligns with how the weights are being loaded. Any architectural discrepancies could prevent the model from producing the expected outputs.
- After loading the weights, verify if all layers are correctly populated by printing summaries or specific weights.
- Check the input image preprocessing steps; sometimes discrepancies in input formatting can lead to unexpected outputs.

Here’s a snippet to check weights for a specific layer after loading:
```python
print("Weights of 21st layer: ", seg_model.model.model[21].state_dict())
```

Also, ensure that after image prediction the data types and structures are compatible with what your visualization and saving functions expect.

If these suggestions don’t resolve the issue, it would be helpful to debug by isolating the step where it fails to carry forward the expected outputs. Keep us posted on your progress!
